classes/task/clear_old_lockouts.php

```diff
@@ -0,0 +1,88 @@
+<?php
+// This file is part of Moodle - http://moodle.org/
+//
+// Moodle is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// Moodle is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with Moodle.  If not, see <http://www.gnu.org/licenses/>.
+
+/**
+ * Task to clear old lockouts.
+ *
+ * @package     tool_securityquestions
+ * @copyright   Peter Burnett <peterburnett@catalyst-au.net>
+ * @license     http://www.gnu.org/copyleft/gpl.html GNU GPL v3 or later
+ */
+
+namespace tool_securityquestions\task;
+defined('MOODLE_INTERNAL') || die();
+// Manually require locallib to find unlock functions.
+require_once(__DIR__ . '/../../locallib.php');
+
+/**
+ * Clear old lockouts task class.
+ *
+ * @package    tool_securityquestions
+ * @copyright  Peter Burnett <peterburnett@catalyst-au.net>
+ * @license    http://www.gnu.org/copyleft/gpl.html GNU GPL v3 or later
+ */
+class clear_old_lockouts extends \core\task\scheduled_task {
+
+    /**
+     * Name getter.
+     *
+     * @return string name of the task.
+     */
+    public function get_name() {
+        return get_string('taskclearoldlockouts', 'tool_securityquestions');
+    }
+
+    /**
+     * Entry point for the task handler.
+     *
+     * @return void
+     */
+    public function execute() {
+        mtrace('Starting to clear old security questions lockouts.');
+        $this->clear_lockouts();
+        mtrace('Finished clearing old security question lockouts.');
+    }
+
+    /**
+     * Performs the unlock of all of the users with old lockouts.
+     *
+     * @return void
+     */
+    private function clear_lockouts() {
+        global $DB;
+
+        $period = get_config('tool_securityquestions', 'lockoutresetduration');
+        // If period is 0, pull out early and don't clear anyone.
+        if ($period === 0) {
+            return;
+        }
+
+        $timetoclear = time() - $period;
+        // Grab all records older than the lockout reset duration, where they are atleast tier 1.
+        $sqlfrag = "timefailed < :time AND tier > 0";
+        $records = $DB->get_records_select('tool_securityquestions_loc', $sqlfrag, array('time' => $timetoclear));
```

Comment:
How about instead of a loc cleanup, instead of just timefailed, we also add a new column which is 'lockeduntil' and calculated at lockout time,

Then we never delete the lockout history and just query if there are any lockouts which are not expired

Reply:
This can instead be done even more cleanly during the lock initialisation check, where any expired locks can be cleared, and the user's lock will be treated as a new lock